pylint/checkers/unicode.py

```diff
@@ -0,0 +1,701 @@
+    codec = codec.replace("-", "").replace(" ", "").lower()
+    if codec.startswith("utf32"):
+        return 4
+    if codec.startswith("utf16"):
+        return 2
+
+    return 1
+
+
+@lru_cache(maxsize=1000)
+def _cached_encode_search(string: str, encoding: str) -> bytes:
+    """A cached version of encode used for search pattern"""
+    return _encode_without_bom(string, encoding)
+
+
+def _fix_utf16_32_line_stream(steam: Iterable[bytes], codec: str) -> Iterable[bytes]:
+    """Handle line ending for UTF16 and UTF32 correctly
+
+    Currently Python simply strips the required zeros after \n after the
+    line ending. Leading to lines that can't be decoded propery
+    """
+    codec = _normalize_codec_name(codec)
+    if not codec.startswith("utf-16") and not codec.startswith("utf-32"):
+        yield from steam
+    else:
+        # UTF 16 and UTF-32 are sucking, so first get all the bytes in memory
+        content = b"".join(line for line in steam)
+
+        new_line = _cached_encode_search("\n", codec)
+        # Now we split the line by the real new line in the correct encoding
+        # we can't use split as it would strip the \n that we need
+
+        start: int = 0
+        pos: int
+        while True:
+            pos = content.find(new_line, start)
+            if pos >= 0:
+                yield content[start : pos + len(new_line)]
+            else:
+                # Yield the rest and finish
+                yield content[start:]
+                break
+
+            start = pos + len(new_line)
+
+
+def extract_codec_from_bom(first_line: bytes) -> Optional[str]:
+    """Try to extract the codec (unicode only) by checking for the BOM
+
+    For details about BOM see https://unicode.org/faq/utf_bom.html#BOM
+
+    Args:
+        first_line: the first line of a file
+
+    Returns:
+        a codec name
+    """
+    for bom, codec in get_bom_sorted().items():
+        if first_line.startswith(bom):
+            return codec
+
+    return None
+
+
+class UnicodeChecker(BaseChecker):
+    """Check characters that could be used to hide bad code to humans
+
+    This includes:
+
+    - Bidirectional Unicode (see https://trojansource.codes/)
+
+    - Bad ASCII characters (see PEP672)
+
+        If a programmer requires to use such a character they should use the escaped
+        version, that is also much easier to read and does not depend on the editor used.
+
+    The Checker also includes a check that UTF-16 and UTF-32 are not used to encode
+    Python files.
+
+    At the time of writing Python supported only UTF-8. See
+    https://stackoverflow.com/questions/69897842/ and https://bugs.python.org/issue1503789
+    for background.
+    """
+
+    __implements__ = interfaces.IRawChecker
+    priority = -1
+
+    name = "unicode_checker"
+
+    # Based on:
+    # https://golangexample.com/go-linter-which-checks-for-dangerous-unicode-character-sequences/
+    _BIDI_UNICODE = [
```

Comment:
```suggestion
    # Based on:
    # https://golangexample.com/go-linter-which-checks-for-dangerous-unicode-character-sequences/
    # We use '\u' because it doesn't require a map lookup and is therefore faster
    _BIDI_UNICODE = [
```